starknet_py/compile/compiler.py

```diff
@@ -0,0 +1,127 @@
+import json
+import os
+from pathlib import Path
+from typing import List, NewType, Optional, Tuple, Union
+
+from starkware.starknet.services.api.contract_definition import ContractDefinition
+from starkware.cairo.lang.compiler.constants import MAIN_SCOPE, LIBS_DIR_ENVVAR
+from starkware.cairo.lang.cairo_constants import DEFAULT_PRIME
+from starkware.cairo.lang.compiler.cairo_compile import (
+    get_module_reader,
+)
+from starkware.cairo.lang.compiler.preprocessor.preprocess_codes import preprocess_codes
+from starkware.starknet.compiler.compile import assemble_starknet_contract
+from starkware.starknet.compiler.starknet_pass_manager import starknet_pass_manager
+
+CairoSourceCode = NewType("CairoSourceCode", str)
+CairoFilename = NewType("CairoFilename", str)
+StarknetCompilationSource = NewType(
+    "CairoSource", Union[CairoSourceCode, List[CairoFilename]]
+)
+
+
+class Compiler:
+    """
+    Class for compiling Cairo contracts
+    """
+
+    @staticmethod
+    def create_contract_definition(
+        compiled_contract: Optional[str] = None,
+        contract_source: Optional[StarknetCompilationSource] = None,
+        search_paths: Optional[List[str]] = None,
+    ) -> ContractDefinition:
+        """
+        Creates ContractDefinition either from already compiled contract or contract source code
+
+        :raises ValueError: when neither contract_source and compiled_contract are given
+        :param compiled_contract: an already compiled contract
+        :param contract_source: string containing source code or a list of source files paths
+        :param search_paths: a ``list`` of paths used by starknet_compile to resolve dependencies within contracts
+        :return: a ContractDefinition
+        """
+        if not contract_source and not compiled_contract:
+            raise ValueError(
+                "One of compiled_contract or compilation_source is required."
+            )
+
+        if not compiled_contract:
+            compiled_contract = Compiler.compile_contract(
```

Comment:
I am curious why we are using static methods for everything? Couldn't these just be normal methods on the class so you can do `self.compile_contract` here? Else, couldn't all these methods just occur at the module level?

Reply:
Yep we should probably instantiate compiler with `compilation_source, search_paths` and then use methods accordingly

Reply:
We can probably change these methods to non static, but I think that `compilation_source` and `search_paths` should still be methods' parameters. Otherwise you'd have to initialize a new compiler instance every time you wanted to compile some other contract